File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/discovery/providers/client/ServiceDiscoveryProviderSpec.kt
package stasis.test.client_android.lib.discovery.providers.client

import io.kotest.assertions.throwables.shouldNotThrow
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.ints.shouldBeGreaterThanOrEqual
import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.types.beInstanceOf
import kotlinx.coroutines.delay
import stasis.client_android.lib.discovery.ServiceApiClient
import stasis.client_android.lib.discovery.ServiceApiEndpoint
import stasis.client_android.lib.discovery.ServiceDiscoveryClient
import stasis.client_android.lib.discovery.ServiceDiscoveryResult
import stasis.client_android.lib.discovery.exceptions.DiscoveryFailure
import stasis.client_android.lib.discovery.providers.client.ServiceDiscoveryProvider
import stasis.client_android.lib.model.core.networking.EndpointAddress
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Failure
import stasis.test.client_android.lib.eventually
import stasis.test.client_android.lib.mocks.MockServiceDiscoveryClient
import java.time.Duration
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

class ServiceDiscoveryProviderSpec : WordSpec({
    "A ServiceDiscoveryProvider" should {

        class TestApiClient : ServiceApiClient

        class TestCoreClient : ServiceApiClient

        fun createClients(
            initialDiscoveryResult: ServiceDiscoveryResult,
            nextDiscoveryResult: ServiceDiscoveryResult
        ): List<ServiceApiClient> = listOf(
            MockServiceDiscoveryClient(
                initialDiscoveryResult = initialDiscoveryResult,
                nextDiscoveryResult = nextDiscoveryResult
            )
        )

        fun createClientFactory(
            nextDiscoveryResult: ServiceDiscoveryResult = ServiceDiscoveryResult.KeepExisting
        ): ServiceApiClient.Factory =
            object : ServiceApiClient.Factory {
                override fun create(endpoint: ServiceApiEndpoint.Api, coreClient: ServiceApiClient): ServiceApiClient =
                    TestApiClient()

                override fun create(endpoint: ServiceApiEndpoint.Core): ServiceApiClient =
                    TestCoreClient()

                override fun create(endpoint: ServiceApiEndpoint.Discovery): ServiceApiClient =
                    MockServiceDiscoveryClient(
                        initialDiscoveryResult = nextDiscoveryResult,
                        nextDiscoveryResult = nextDiscoveryResult
                    )
            }

        suspend fun <T> managedProvider(provider: ServiceDiscoveryProvider, block: suspend () -> T): T =
            try {
                block()
            } finally {
                provider.stop()
            }

        "support providing existing clients when discovery is not active" {
            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofSeconds(3),
                interval = Duration.ofSeconds(3),
                initialClients = createClients(
                    initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
                    nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
                ),
                clientFactory = createClientFactory(),
                coroutineScope = testScope
            ).get()

            managedProvider(provider) {
                provider should beInstanceOf<ServiceDiscoveryProvider.Disabled>()

                provider.latest(ServiceDiscoveryClient::class.java) should beInstanceOf<MockServiceDiscoveryClient>()

                shouldThrow<DiscoveryFailure> { provider.latest(TestApiClient::class.java) }
                shouldThrow<DiscoveryFailure> { provider.latest(TestCoreClient::class.java) }
            }
        }

        "support providing new clients when discovery is active" {
            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofSeconds(3),
                interval = Duration.ofSeconds(3),
                initialClients = createClients(
                    initialDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
                        endpoints = ServiceDiscoveryResult.Endpoints(
                            api = ServiceApiEndpoint.Api(uri = "test-rui"),
                            core = ServiceApiEndpoint.Core(address = EndpointAddress.HttpEndpointAddress(uri = "test-rui")),
                            discovery = ServiceApiEndpoint.Discovery(uri = "test-rui")
                        ),
                        recreateExisting = false
                    ),
                    nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
                ),
                clientFactory = createClientFactory(),
                coroutineScope = testScope
            ).get()

            managedProvider(provider) {
                provider should beInstanceOf<ServiceDiscoveryProvider.Default>()

                provider.latest(ServiceDiscoveryClient::class.java) should beInstanceOf<MockServiceDiscoveryClient>()

                shouldNotThrow<Throwable> { provider.latest(TestApiClient::class.java) }
                shouldNotThrow<Throwable> { provider.latest(TestCoreClient::class.java) }
            }
        }

        "periodically refresh list of endpoints and clients (with result=keep-existing)" {
            fun createResult(recreateExisting: Boolean = false): ServiceDiscoveryResult =
                ServiceDiscoveryResult.SwitchTo(
                    endpoints = ServiceDiscoveryResult.Endpoints(
                        api = ServiceApiEndpoint.Api(uri = java.util.UUID.randomUUID().toString()),
                        core = ServiceApiEndpoint.Core(
                            address = EndpointAddress.HttpEndpointAddress(
                                uri = java.util.UUID.randomUUID().toString()
                            )
                        ),
                        discovery = ServiceApiEndpoint.Discovery(uri = java.util.UUID.randomUUID().toString())
                    ),
                    recreateExisting = recreateExisting
                )

            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofMillis(200),
                interval = Duration.ofMillis(200),
                initialClients = createClients(
                    initialDiscoveryResult = createResult(),
                    nextDiscoveryResult = createResult()
                ),
                clientFactory = createClientFactory(),
                coroutineScope = testScope
            ).get()

            managedProvider(provider) {
                provider should beInstanceOf<ServiceDiscoveryProvider.Default>()

                val initialDiscoveryClient = provider.latest(ServiceDiscoveryClient::class.java)
                val initialApiClient = provider.latest(TestApiClient::class.java)
                val initialCoreClient = provider.latest(TestCoreClient::class.java)

                delay(timeMillis = 100)

                provider.latest(ServiceDiscoveryClient::class.java) shouldBe (initialDiscoveryClient)
                provider.latest(TestApiClient::class.java) shouldBe (initialApiClient)
                provider.latest(TestCoreClient::class.java) shouldBe (initialCoreClient)
            }
        }

        "periodically refresh list of endpoints and clients (with result=switch-to)" {
            fun createResult(recreateExisting: Boolean = false): ServiceDiscoveryResult =
                ServiceDiscoveryResult.SwitchTo(
                    endpoints = ServiceDiscoveryResult.Endpoints(
                        api = ServiceApiEndpoint.Api(uri = java.util.UUID.randomUUID().toString()),
                        core = ServiceApiEndpoint.Core(
                            address = EndpointAddress.HttpEndpointAddress(
                                uri = java.util.UUID.randomUUID().toString()
                            )
                        ),
                        discovery = ServiceApiEndpoint.Discovery(uri = java.util.UUID.randomUUID().toString())
                    ),
                    recreateExisting = recreateExisting
                )

            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofMillis(200),
                interval = Duration.ofMillis(200),
                initialClients = createClients(
                    initialDiscoveryResult = createResult(),
                    nextDiscoveryResult = createResult()
                ),
                clientFactory = createClientFactory(nextDiscoveryResult = createResult(recreateExisting = true)),
                coroutineScope = testScope
            ).get()

            managedProvider(provider) {
                provider should beInstanceOf<ServiceDiscoveryProvider.Default>()

                val initialDiscoveryClient = provider.latest(ServiceDiscoveryClient::class.java)
                val initialApiClient = provider.latest(TestApiClient::class.java)
                val initialCoreClient = provider.latest(TestCoreClient::class.java)

                delay(timeMillis = 100)

                provider.latest(ServiceDiscoveryClient::class.java) shouldBe (initialDiscoveryClient)
                provider.latest(TestApiClient::class.java) shouldBe (initialApiClient)
                provider.latest(TestCoreClient::class.java) shouldBe (initialCoreClient)

                eventually {
                    provider.latest(ServiceDiscoveryClient::class.java) shouldNotBe (initialDiscoveryClient)
                    provider.latest(TestApiClient::class.java) shouldNotBe (initialApiClient)
                    provider.latest(TestCoreClient::class.java) shouldNotBe (initialCoreClient)
                }
            }
        }

        "not recreate clients for same endpoints" {
            fun createResult(recreateExisting: Boolean = false): ServiceDiscoveryResult =
                ServiceDiscoveryResult.SwitchTo(
                    endpoints = ServiceDiscoveryResult.Endpoints(
                        api = ServiceApiEndpoint.Api(uri = java.util.UUID.randomUUID().toString()),
                        core = ServiceApiEndpoint.Core(
                            address = EndpointAddress.HttpEndpointAddress(
                                uri = java.util.UUID.randomUUID().toString()
                            )
                        ),
                        discovery = ServiceApiEndpoint.Discovery(uri = java.util.UUID.randomUUID().toString())
                    ),
                    recreateExisting = recreateExisting
                )

            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofMillis(100),
                interval = Duration.ofMillis(100),
                initialClients = createClients(
                    initialDiscoveryResult = createResult(),
                    nextDiscoveryResult = createResult()
                ),
                clientFactory = createClientFactory(nextDiscoveryResult = createResult()),
                coroutineScope = testScope
            ).get()

            managedProvider(provider) {
                provider should beInstanceOf<ServiceDiscoveryProvider.Default>()

                val initialDiscoveryClient = provider.latest(ServiceDiscoveryClient::class.java)
                val initialApiClient = provider.latest(TestApiClient::class.java)
                val initialCoreClient = provider.latest(TestCoreClient::class.java)

                delay(timeMillis = 200)

                provider.latest(ServiceDiscoveryClient::class.java) shouldNotBe (initialDiscoveryClient)
                provider.latest(TestApiClient::class.java) shouldNotBe (initialApiClient)
                provider.latest(TestCoreClient::class.java) shouldNotBe (initialCoreClient)

                val latestDiscoveryClient = provider.latest(ServiceDiscoveryClient::class.java)
                val latestApiClient = provider.latest(TestApiClient::class.java)
                val latestCoreClient = provider.latest(TestCoreClient::class.java)

                delay(timeMillis = 200)

                eventually {
                    latestDiscoveryClient shouldNotBe initialDiscoveryClient
                    latestApiClient shouldNotBe initialApiClient
                    latestCoreClient shouldNotBe initialCoreClient

                    provider.latest(ServiceDiscoveryClient::class.java) shouldBe (latestDiscoveryClient)
                    provider.latest(TestApiClient::class.java) shouldBe (latestApiClient)
                    provider.latest(TestCoreClient::class.java) shouldBe (latestCoreClient)
                }
            }
        }

        "fail to retrieve unsupported client types" {
            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofSeconds(3),
                interval = Duration.ofSeconds(3),
                initialClients = createClients(
                    initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
                    nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
                ),
                clientFactory = createClientFactory(),
                coroutineScope = testScope
            ).get()


            managedProvider(provider) {
                provider should beInstanceOf<ServiceDiscoveryProvider.Disabled>()

                val e = shouldThrow<DiscoveryFailure> { provider.latest(TestApiClient::class.java) }

                e.message shouldBe ("Service client [${TestApiClient::class.java.name}] was not found")
            }
        }

        "handle discovery failures" {
            val clientCalls = AtomicInteger(0)

            val provider = ServiceDiscoveryProvider(
                initialDelay = Duration.ofMillis(100),
                interval = Duration.ofMillis(200),
                initialClients = createClients(
                    initialDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
                        endpoints = ServiceDiscoveryResult.Endpoints(
                            api = ServiceApiEndpoint.Api(uri = "test-rui"),
                            core = ServiceApiEndpoint.Core(address = EndpointAddress.HttpEndpointAddress(uri = "test-rui")),
                            discovery = ServiceApiEndpoint.Discovery(uri = "test-rui")
                        ),
                        recreateExisting = false
                    ),
                    nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
                ),
                clientFactory = object : ServiceApiClient.Factory {
                    override fun create(
                        endpoint: ServiceApiEndpoint.Api,
                        coreClient: ServiceApiClient
                    ): ServiceApiClient =
                        TestApiClient()

                    override fun create(endpoint: ServiceApiEndpoint.Core): ServiceApiClient =
                        TestCoreClient()

                    override fun create(endpoint: ServiceApiEndpoint.Discovery): ServiceApiClient =
                        object : ServiceDiscoveryClient {
                            override val attributes: ServiceDiscoveryClient.Attributes =
                                MockServiceDiscoveryClient.TestAttributes(a = "b")

                            override suspend fun latest(isInitialRequest: Boolean): Try<ServiceDiscoveryResult> {
                                clientCalls.incrementAndGet()
                                return Failure(RuntimeException("Test failure"))
                            }
                        }
                },
                coroutineScope = testScope
            ).get()

            managedProvider(provider) {
                clientCalls.get() shouldBe (0)

                delay(timeMillis = 100)

                clientCalls.get() shouldBe (0)

                delay(timeMillis = 400)

                clientCalls.get() shouldBeGreaterThanOrEqual (3) // the interval is reduced so more requests should be made
            }
        }

        "support creating the provider asynchronously" {
            val providerRef = AtomicReference<Try<ServiceDiscoveryProvider>?>(null)

            ServiceDiscoveryProvider.create(
                initialDelay = Duration.ofSeconds(3),
                interval = Duration.ofSeconds(3),
                initialClients = createClients(
                    initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
                    nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
                ),
                clientFactory = createClientFactory(),
                onCreated = { providerRef.set(it) },
                coroutineScope = testScope
            )

            providerRef.get() shouldBe (null)

            eventually {
                val provider = providerRef.get()?.get()
                provider shouldNotBe (null)

                managedProvider(provider!!) {
                    provider should beInstanceOf<ServiceDiscoveryProvider.Disabled>()
                }
            }
        }
    }
})

File: core/src/test/scala/stasis/test/specs/unit/core/discovery/providers/client/ServiceDiscoveryProviderSpec.scala
package stasis.test.specs.unit.core.discovery.providers.client

import java.util.concurrent.atomic.AtomicInteger

import scala.concurrent.Future
import scala.concurrent.duration._

import org.apache.pekko.actor.typed.ActorSystem
import org.apache.pekko.actor.typed.scaladsl.Behaviors
import org.scalatest.Assertion
import org.scalatest.BeforeAndAfterAll
import org.scalatest.concurrent.Eventually

import stasis.core.discovery.ServiceApiClient
import stasis.core.discovery.ServiceApiEndpoint
import stasis.core.discovery.ServiceDiscoveryClient
import stasis.core.discovery.ServiceDiscoveryResult
import stasis.core.discovery.exceptions.DiscoveryFailure
import stasis.core.discovery.providers.client.ServiceDiscoveryProvider
import stasis.core.networking.http.HttpEndpointAddress
import io.github.sndnv.layers.testing.UnitSpec
import stasis.test.specs.unit.core.discovery.mocks.MockServiceDiscoveryClient
import stasis.test.specs.unit.core.discovery.providers.client.ServiceDiscoveryProviderSpec.TestApiClient
import stasis.test.specs.unit.core.discovery.providers.client.ServiceDiscoveryProviderSpec.TestCoreClient

class ServiceDiscoveryProviderSpec extends UnitSpec with Eventually with BeforeAndAfterAll {
  "A ServiceDiscoveryProvider" should "support providing existing clients when discovery is not active" in {
    val provider = ServiceDiscoveryProvider(
      interval = 3.seconds,
      initialClients = createClients(
        initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
        nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
      ),
      clientFactory = createClientFactory()
    ).await

    provider should be(a[ServiceDiscoveryProvider.Disabled])

    provider.latest[ServiceDiscoveryClient] should be(a[MockServiceDiscoveryClient])
    a[DiscoveryFailure] should be thrownBy provider.latest[TestApiClient]
    a[DiscoveryFailure] should be thrownBy provider.latest[TestCoreClient]
  }

  it should "support providing new clients when discovery is active" in {
    val provider = ServiceDiscoveryProvider(
      interval = 3.seconds,
      initialClients = createClients(
        initialDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
          endpoints = ServiceDiscoveryResult.Endpoints(
            api = ServiceApiEndpoint.Api(uri = "test-rui"),
            core = ServiceApiEndpoint.Core(address = HttpEndpointAddress(uri = "test-rui")),
            discovery = ServiceApiEndpoint.Discovery(uri = "test-rui")
          ),
          recreateExisting = false
        ),
        nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
      ),
      clientFactory = createClientFactory()
    ).await

    provider should be(a[ServiceDiscoveryProvider.Default])

    provider.latest[ServiceDiscoveryClient] should be(a[MockServiceDiscoveryClient])
    noException should be thrownBy provider.latest[TestApiClient]
    noException should be thrownBy provider.latest[TestCoreClient]
  }

  it should "periodically refresh list of endpoints and clients (with result=keep-existing)" in {
    def createResult(recreateExisting: Boolean = false): ServiceDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
      endpoints = ServiceDiscoveryResult.Endpoints(
        api = ServiceApiEndpoint.Api(uri = java.util.UUID.randomUUID().toString),
        core = ServiceApiEndpoint.Core(address = HttpEndpointAddress(uri = java.util.UUID.randomUUID().toString)),
        discovery = ServiceApiEndpoint.Discovery(uri = java.util.UUID.randomUUID().toString)
      ),
      recreateExisting = recreateExisting
    )

    val provider = ServiceDiscoveryProvider(
      interval = 200.millis,
      initialClients = createClients(initialDiscoveryResult = createResult(), nextDiscoveryResult = createResult()),
      clientFactory = createClientFactory()
    ).await

    provider should be(a[ServiceDiscoveryProvider.Default])

    val initialDiscoveryClient = provider.latest[ServiceDiscoveryClient]
    val initialApiClient = provider.latest[TestApiClient]
    val initialCoreClient = provider.latest[TestCoreClient]

    await(delay = 100.millis)

    provider.latest[ServiceDiscoveryClient] should be(initialDiscoveryClient)
    provider.latest[TestApiClient] should be(initialApiClient)
    provider.latest[TestCoreClient] should be(initialCoreClient)
  }

  it should "periodically refresh list of endpoints and clients (with result=switch-to)" in {
    def createResult(recreateExisting: Boolean = false): ServiceDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
      endpoints = ServiceDiscoveryResult.Endpoints(
        api = ServiceApiEndpoint.Api(uri = java.util.UUID.randomUUID().toString),
        core = ServiceApiEndpoint.Core(address = HttpEndpointAddress(uri = java.util.UUID.randomUUID().toString)),
        discovery = ServiceApiEndpoint.Discovery(uri = java.util.UUID.randomUUID().toString)
      ),
      recreateExisting = recreateExisting
    )

    val provider = ServiceDiscoveryProvider(
      interval = 200.millis,
      initialClients = createClients(initialDiscoveryResult = createResult(), nextDiscoveryResult = createResult()),
      clientFactory = createClientFactory(nextDiscoveryResult = createResult(recreateExisting = true))
    ).await

    provider should be(a[ServiceDiscoveryProvider.Default])

    val initialDiscoveryClient = provider.latest[ServiceDiscoveryClient]
    val initialApiClient = provider.latest[TestApiClient]
    val initialCoreClient = provider.latest[TestCoreClient]

    await(delay = 100.millis)

    provider.latest[ServiceDiscoveryClient] should be(initialDiscoveryClient)
    provider.latest[TestApiClient] should be(initialApiClient)
    provider.latest[TestCoreClient] should be(initialCoreClient)

    eventually[Assertion] {
      provider.latest[ServiceDiscoveryClient] should not be initialDiscoveryClient
      provider.latest[TestApiClient] should not be initialApiClient
      provider.latest[TestCoreClient] should not be initialCoreClient
    }
  }

  it should "not recreate clients for same endpoints" in {
    def createResult(): ServiceDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
      endpoints = ServiceDiscoveryResult.Endpoints(
        api = ServiceApiEndpoint.Api(uri = java.util.UUID.randomUUID().toString),
        core = ServiceApiEndpoint.Core(address = HttpEndpointAddress(uri = java.util.UUID.randomUUID().toString)),
        discovery = ServiceApiEndpoint.Discovery(uri = java.util.UUID.randomUUID().toString)
      ),
      recreateExisting = false
    )

    val provider = ServiceDiscoveryProvider(
      interval = 100.millis,
      initialClients = createClients(initialDiscoveryResult = createResult(), nextDiscoveryResult = createResult()),
      clientFactory = createClientFactory(nextDiscoveryResult = createResult())
    ).await

    provider should be(a[ServiceDiscoveryProvider.Default])

    val initialDiscoveryClient = provider.latest[ServiceDiscoveryClient]
    val initialApiClient = provider.latest[TestApiClient]
    val initialCoreClient = provider.latest[TestCoreClient]

    await(delay = 200.millis)

    provider.latest[ServiceDiscoveryClient] should not be initialDiscoveryClient
    provider.latest[TestApiClient] should not be initialApiClient
    provider.latest[TestCoreClient] should not be initialCoreClient

    val latestDiscoveryClient = provider.latest[ServiceDiscoveryClient]
    val latestApiClient = provider.latest[TestApiClient]
    val latestCoreClient = provider.latest[TestCoreClient]

    await(delay = 200.millis)

    eventually[Assertion] {
      latestDiscoveryClient should not be initialDiscoveryClient
      latestApiClient should not be initialApiClient
      latestCoreClient should not be initialCoreClient

      provider.latest[ServiceDiscoveryClient] should be(latestDiscoveryClient)
      provider.latest[TestApiClient] should be(latestApiClient)
      provider.latest[TestCoreClient] should be(latestCoreClient)
    }
  }

  it should "fail to retrieve unsupported client types" in {
    val provider = ServiceDiscoveryProvider(
      interval = 3.seconds,
      initialClients = createClients(
        initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
        nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
      ),
      clientFactory = createClientFactory()
    ).await

    provider should be(a[ServiceDiscoveryProvider.Disabled])

    val e = intercept[DiscoveryFailure](provider.latest[TestApiClient])
    e.getMessage should be(s"Service client [${classOf[TestApiClient].getName}] was not found")
  }

  it should "handle discovery failures" in {
    val clientCalls = new AtomicInteger(0)

    val _ = ServiceDiscoveryProvider(
      interval = 200.millis,
      initialClients = createClients(
        initialDiscoveryResult = ServiceDiscoveryResult.SwitchTo(
          endpoints = ServiceDiscoveryResult.Endpoints(
            api = ServiceApiEndpoint.Api(uri = "test-rui"),
            core = ServiceApiEndpoint.Core(address = HttpEndpointAddress(uri = "test-rui")),
            discovery = ServiceApiEndpoint.Discovery(uri = "test-rui")
          ),
          recreateExisting = false
        ),
        nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
      ),
      clientFactory = new ServiceApiClient.Factory {
        override def create(endpoint: ServiceApiEndpoint.Api, coreClient: ServiceApiClient): ServiceApiClient =
          new TestApiClient {}

        override def create(endpoint: ServiceApiEndpoint.Core): ServiceApiClient =
          new TestCoreClient {}

        override def create(endpoint: ServiceApiEndpoint.Discovery): ServiceApiClient =
          new ServiceDiscoveryClient {
            override val attributes: ServiceDiscoveryClient.Attributes =
              MockServiceDiscoveryClient.TestAttributes(a = "b")

            override def latest(isInitialRequest: Boolean): Future[ServiceDiscoveryResult] = {
              val _ = clientCalls.incrementAndGet()
              Future.failed(new RuntimeException("Test failure"))
            }
          }
      }
    ).await

    clientCalls.get() should be(0)

    await(delay = 100.millis)

    clientCalls.get() should be(0)

    await(delay = 300.millis)

    eventually {
      clientCalls.get() should be >= 5 // the interval is reduced so more requests should be made
    }
  }

  private def createClients(
    initialDiscoveryResult: ServiceDiscoveryResult,
    nextDiscoveryResult: ServiceDiscoveryResult
  ): Seq[ServiceApiClient] = Seq(
    new MockServiceDiscoveryClient(
      initialDiscoveryResult = initialDiscoveryResult,
      nextDiscoveryResult = nextDiscoveryResult
    )
  )

  private def createClientFactory(
    nextDiscoveryResult: ServiceDiscoveryResult = ServiceDiscoveryResult.KeepExisting
  ): ServiceApiClient.Factory =
    new ServiceApiClient.Factory {
      override def create(endpoint: ServiceApiEndpoint.Api, coreClient: ServiceApiClient): ServiceApiClient =
        new TestApiClient {}

      override def create(endpoint: ServiceApiEndpoint.Core): ServiceApiClient =
        new TestCoreClient {}

      override def create(endpoint: ServiceApiEndpoint.Discovery): ServiceApiClient =
        new MockServiceDiscoveryClient(
          initialDiscoveryResult = nextDiscoveryResult,
          nextDiscoveryResult = nextDiscoveryResult
        )
    }

  override implicit val patienceConfig: PatienceConfig = PatienceConfig(5.seconds, 250.milliseconds)

  private implicit val typedSystem: ActorSystem[Nothing] = ActorSystem(
    guardianBehavior = Behaviors.ignore,
    name = "ServiceDiscoveryProviderSpec"
  )

  override protected def afterAll(): Unit = typedSystem.terminate()
}

object ServiceDiscoveryProviderSpec {
  class TestApiClient extends ServiceApiClient
  class TestCoreClient extends ServiceApiClient
}
